lib/seahorn/Houdini.cc

```diff
@@ -0,0 +1,292 @@
+
+		  //apply head_cand to actual arguments
+		  ExprMap head_bvar_map;
+		  //I need a visitor here.
+		  ExprVector bvars;
+		  get_all_bvars(head_cand, std::back_inserter(bvars));
+
+		  for(ExprVector::iterator it = bvars.begin(); it != bvars.end(); ++it)
+		  {
+			  //outs() << *(*it) << "\n";
+			  unsigned bvar_id = bind::bvarId(*it);
+			  Expr head_app_arg = bind::domainTy(head_app, bvar_id);// To improve
+			  head_bvar_map.insert(std::pair<Expr,Expr>(*it, head_app_arg));
+		  }
+		  //outs() << "[HEAD BVAR MAP]:\n";
+		  //outs() << "[MAP SIZE]: " << head_bvar_map.size() << "\n";
+		  //for (ExprMap::iterator it = head_bvar_map.begin(); it!=head_bvar_map.end(); ++it)
+		  //{
+		  //  outs() << "[KEY]: " << *(it->first) << ", [VALUE]: " << *(it->second) << "\n";
+		  //}
+
+		  Expr head_cand_app = replace(head_cand, head_bvar_map);
+
+		  //cand_app is application of the overall conjunction
+		  Expr neg_head_cand_app = mk<NEG>(head_cand_app);
+
+		  //outs() << "[NEG HEAD CAND APP]: " << *neg_head_cand_app << "\n";
+
+		  //generate candidate for rule body
+		  Expr body = r.body();
+		  ExprMap body_map;
+		  ExprVector body_pred_app_vec;
+		  get_all_pred_apps(body, db, std::back_inserter(body_pred_app_vec));
+		  //outs() << "[BODY PRED NUM]: " << body_pred_app_vec.size() << "\n";
+		  for (ExprVector::iterator it = body_pred_app_vec.begin(); it != body_pred_app_vec.end(); ++it)
+		  {
+			  Expr body_pred_app = *it;
+			  Expr body_pred = bind::fname(body_pred_app);
+			  Expr body_cand = guessCandidate(body_pred);
+			  //outs() << "[BODY CAND]: " << *body_cand << "\n";
+			  //apply body_cand to actual arguments
+			  ExprMap body_bvar_map;
+			  ExprVector body_bvars;
+			  get_all_bvars(body_cand, std::back_inserter(body_bvars));
+			  for(ExprVector::iterator it_bv = body_bvars.begin(); it_bv != body_bvars.end(); ++it_bv)
+			  {
+				  unsigned bvar_id = bind::bvarId(*it_bv);
+				  Expr body_app_arg = bind::domainTy(body_pred_app, bvar_id);// To improve
+				  body_bvar_map.insert(std::pair<Expr,Expr>(*it_bv, body_app_arg));
+			  }
+			  Expr body_cand_app = replace(body_cand, body_bvar_map);
+			  //outs() << "[BODY CAND APP]: " << *body_cand_app << "\n";
+			  body_map.insert(std::pair<Expr, Expr>(body_pred_app, body_cand_app));
+		  }
+		  Expr new_body = replace(body, body_map);
+
+		  //outs() << "[NEW BODY]: " << *new_body << "\n";
+
+		  //update cand and cand_app
+		  Expr cand_app = mk<AND>(neg_head_cand_app, new_body);
+
+		  outs() << "[CANDIDATE]: " << *cand_app << "\n";
+
+		  if(validateRule(cand_app, hm) == SAT)
+		  {
+			 //add rules in db.use(r.head()) to worklist
+			 Expr head_app = r.head();
+			 Expr head_fdecl = bind::fname(head_app);
+			 outs() << "[USE SET SIZE]: " << db.use(head_fdecl).size() << "\n";
+			 for(auto r_use : db.use(head_fdecl))
+		     {
+				 bool isExist = false;
+				 for(HornClauseDB::RuleVector::iterator itr=workList.begin(); itr!=workList.end(); ++itr)
```

Comment:
Use STL instead of iterating yourself.

This piece of code suggests that worklist is not indexed correctly. Adding / removing elements from the worklist should be efficient.